### ISSUE TITLE
Blazor worker "memory access out of bounds" bug

### ISSUE THREAD

**Issue by StachuDotNet** [darklang/dark#4059]
![image](https://user-images.githubusercontent.com/906686/171303014-055015a1-df07-407d-a870-5f8c6d023dc6.png)

seems to happen only with very large requests.

- https://stackoverflow.com/questions/57938081/blazor-preview-9-mono-wasm-memory-access-out-of-bounds-max-string-size-for-dotn
- https://www.reddit.com/r/WebAssembly/comments/g44ldd/what_are_some_possible_causes_of_runtimeerror/
- https://github.com/dotnet/aspnetcore/issues/5617

**Comment by StachuDotNet** [comment]
I've now gotten the analysis request payload that causes the issue.
It's indeed quite large (70814 chars of JSON).

FWIW this request seems to include a lot of stuff that it doesn't need to perform the analysis.

I'm seeing three paths forward:
- somehow increase the memory limit
- send less stuff in the request (i.e. filter out stuff not needed for analysis)
- send the request in batches, e.g. per https://stackoverflow.com/a/57973792

Currently experimenting to see what the char threshold of failure is.

**Comment by StachuDotNet** [comment]
Note: I can replicate this in Release AOT builds, but not Debug builds.
Now checking to see if I can repro in a non-AOT release build.

**Comment by StachuDotNet** [comment]
Seems I can't - i.e. the bug seems AOT-specific.

**Comment by pbiggar** [comment]
Do you know what function this is happening in? Perhaps we're able to separate reading the json from doing the analysis in some way. Maybe we parse the json in JS code and then call wasm using `JSInterop` or whatever it's called.

**Comment by pbiggar** [comment]
Yeah, either of those sound good. I like the ArrayBuffer/file approach but does that also have size limits?

Are we sure this is the problem though? 76k isn't small but it's also not huge. The Stackoverflow thing says the limit is over 50MB now.

**Comment by StachuDotNet** [comment]
I'm now a bit less confident.

I added this to Wasm.fs:

```
  static member CanWeGetHere(message : string) =
    message |> String.length |> printfn "Yes %i"
```

and posted the giant json to _that_ handler rather than the OnMessage one. It works?

**Comment by StachuDotNet** [comment]
I was wrong before! We _are_ getting to within OnMessage. We're even deserializing OK. The failure seems to be happening within the actual analysis.

**Comment by StachuDotNet** [comment]
Even more interesting:

with my code like this:

```
 let! result =
        task {
          match args with
          | Error e -> return Error e
          | Ok args ->
            try
              printfn "Performing analysis"
              let! result = Eval.performAnalysis args
              printfn "Performed analysis"
              return Ok result
            with
            | e ->
              let metadata = Exception.nestedMetadata e
              System.Console.WriteLine("Error running analysis in Blazor")
              System.Console.WriteLine($"called with message: {message}")
              System.Console.WriteLine(
                $"caught exception: \"{e.Message}\" \"{metadata}\""
              )
              return Error($"exception: {e.Message}, metadata: {metadata}")
        }

      printfn "Serializing result"
      let serialized =
        try
          Json.Vanilla.serialize result
        with
        | e ->
          let metadata = Exception.nestedMetadata e
          System.Console.WriteLine("Error serializing results of Blazor analysis")
          System.Console.WriteLine($"called with message: {message}")
          System.Console.WriteLine(
            $"caught exception: \"{e.Message}\" \"{metadata}\""
          )
          Json.Vanilla.serialize ($"exception: {e.Message}, metadata: {metadata}")

      printfn "Posting result back"
      EvalWorker.postMessage serialized
```

I got "performed analysis" and it dies right after, without ever emitting `Serializing result`. Maybe something to do with the task.

**Comment by pbiggar** [comment]
Well that's optimistic at least!

**Comment by StachuDotNet** [comment]
So, I got it to sort-of work

here's some code:
```
let serializeAndPost result =
        let serialized =
          try
            Json.Vanilla.serialize result
          with
          | e ->
            let metadata = Exception.nestedMetadata e
            System.Console.WriteLine("Error serializing results of Blazor analysis")
            System.Console.WriteLine($"called with message: {message}")
            System.Console.WriteLine(
              $"caught exception: \"{e.Message}\" \"{metadata}\""
            )
            Json.Vanilla.serialize ($"exception: {e.Message}, metadata: {metadata}")

        EvalWorker.postMessage serialized

      match args with
      | Error e -> Error e |> serializeAndPost
      | Ok args ->
        try
          printfn "Performing analysis"
          let! result = Eval.performAnalysis args
          printfn "Performed analysis"
          printfn "%A" result
          Ok result |> serializeAndPost
        with
        | e ->
          let metadata = Exception.nestedMetadata e
          System.Console.WriteLine("Error running analysis in Blazor")
          System.Console.WriteLine($"called with message: {message}")
          System.Console.WriteLine(
            $"caught exception: \"{e.Message}\" \"{metadata}\""
          )
          Error($"exception: {e.Message}, metadata: {metadata}")
          |> serializeAndPost
```

Right now, it successfully runs the full deserialization, analysis, serialization of result, and post-back, and _then_ fails with the memory exception (after js gets the response)

I think the key thing here is that the `task` containing the `result` is garbage-collected after the post-back.
Seems relevant to https://github.com/dotnet/aspnetcore/issues/16692, a closed issue linked from https://github.com/dotnet/aspnetcore/issues/5617

**Comment by StachuDotNet** [comment]
Curiously, if I
- remove the `task{}`
- update the `let! result = Eval.performAnalysis args` to a `let result = Eval.performAnalysis args |> Async.AwaitTask |> Async.RunSynchronously`
I never see `Performed analysis`

Seems to indicate that this is a matter of GC-ing `result` - as opposed to GC-ing `message` (the big incoming payload).
(or maybe it's some combination of the two)

Some notes about this test:
- the incoming (serialized) payload is `593969` chars long
- the response payload (after serialization) is `93622` chars long

**Comment by StachuDotNet** [comment]
Given ^, thinking how to proceed... we can get the analysis results, but it fails upon GC (or something - whenever we exit the `task`). And is unrelated to serializing those results or posting them back to JS.

So, the strat from #4008 seems irrelevant here, for now.

**Comment by StachuDotNet** [comment]
Note: if I...
```
let! result = Eval.performAnalysis args
printfn "Performed analysis"
GC.Collect()
printfn "manually gc'd"
```
We don't reach "manually gc'd"

**Comment by StachuDotNet** [comment]
I queried various methods on `GC` right _before_ and right _after_ performing analysis, with two expressions (2+3, and the _big_ analysis request).

|                Syntax               | 2+3, before | 2 +3, after | big, before | big, after |
|-------------------------------------|-------------|-------------|-------------|------------|
| GetAllocatedBytesForCurrentThread() | 2408736L    | 5665920L    | 15621904L   | 20014232L  |
| GetTotalAllocatedBytes()            | 0L          | 4412192L    | 8888560L    | 17691712L  |
| GetTotalMemory(false)               | 4352064L    | 4918256L    | 13946776L   | 14602064L  |

I haven't actually studied these numbers yet - just collected. Looking now :eyes:

**Comment by StachuDotNet** [comment]
quick thought: maybe we're GCing the analysis engine or something, while we could keep it running eternally instead? Not sure if that makes any sense :)

**Comment by StachuDotNet** [comment]
I wonder if I can replicate this in Debug mode, with a manual `GC.Collect()` - maybe Release mode GCs more often, and the manual call will allow us to get a stack trace? :construction: :crossed_fingers:

**Comment by StachuDotNet** [comment]
A fix has been found (will link PR shortly).
The exact reason the fix works is unfortunately unknown - refer to the upcoming PR for further commentary.

**Comment by StachuDotNet** [comment]
Unfortunately, while #4070 got past that error, another related one occurs shortly after:
```
Uncaught Error: System.ArgumentNullException: ArgumentNull_Generic Arg_ParamName_Name, key
   at System.Collections.Generic.Dictionary`2[[System.Object, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.Int32, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].Remove(Object )
   at System.Runtime.InteropServices.JavaScript.Runtime.ReleaseJSOwnedObjectByGCHandle(Int32 gcHandle)
    at Object._convert_exception_for_method_call (https://static.darklang.com/blazor/dotnet.6.0.5.epfcsc29vy-2e5c0b99229188abd29a4572dbc1760c0a73e6bb6a05b5358b8be034a503a2f0.js:1:176256)
    at Object._handle_exception_for_call (https://static.darklang.com/blazor/dotnet.6.0.5.epfcsc29vy-2e5c0b99229188abd29a4572dbc1760c0a73e6bb6a05b5358b8be034a503a2f0.js:1:178225)
    at Object.managed_BINDINGS_ReleaseJSOwnedObjectByGCHandle [as _release_js_owned_object_by_gc_handle] (https://mono-wasm.invalid/managed_BINDINGS_ReleaseJSOwnedObjectByGCHandle:17:21)
    at Object._js_owned_object_finalized (https://static.darklang.com/blazor/dotnet.6.0.5.epfcsc29vy-2e5c0b99229188abd29a4572dbc1760c0a73e6bb6a05b5358b8be034a503a2f0.js:1:150045)
    at FinalizationRegistry.cleanupSome (<anonymous>)
```

**Comment by pbiggar** [comment]
I believe STJ has a default recusion depth, maybe 250. You can change it.

**Comment by pbiggar** [comment]
https://docs.microsoft.com/en-us/dotnet/api/system.text.json.jsonserializeroptions.maxdepth?view=net-6.0

**Comment by StachuDotNet** [comment]
in case it doesn't respect MaxValue, I'll set it to some reasonably large values smaller than such.
Still making good progress thinning the payload down, fwiw. Getting close to a minimal payload - I can feel it.

**Comment by StachuDotNet** [comment]
Here's the current state of a payload that deserializes OK in Debug or non-AOT Release builds, but not AOT builds (in which, it fails with `memory access out of bounds`):
https://gist.github.com/StachuDotNet/17203813e61fcdfcb2c49c6d9fcf41cf

This started off as user data :sweat_smile: 

Really seems related to depth, but haven't quite figured out how.

**Comment by StachuDotNet** [comment]
When I see this AST, I interpret it as:
```
let x = null // (repeated 16 times)
let x = null
// for a total of 16 bindings for `x`
let x = null
...
```
and then I'm not sure how to read it. Certainly there's at least one pipe involved, but I'm not sure how to grok the fact that there are ... 3 pipes but only one target?

I realize that the AST looks a bit absurd as-is, but: Debug mode is OK deserializing it, and Release mode isn't. I'm trying to figure out why that is. @pbiggar any ideas?

**Comment by StachuDotNet** [comment]
to be clear, if I remove any `let`s (even just one), serialization is OK in AOT mode.

**Comment by StachuDotNet** [comment]
I'm not sure if this is purely a matter of depth (I'll check by replacing a few layers with more `let`s), or something specific to these expression types.

**Comment by pbiggar** [comment]
I'm thinking of what could cause memory errors in this code.

Is there an assertion hit by weird pipes perhaps?

Or a array index?

**Comment by pbiggar** [comment]
What about the pipes? Does removing them work?

**Comment by StachuDotNet** [comment]
Aha! I can remove one of them and still get the error, but if I remove 2 (leaving only 1), no error!

**Comment by StachuDotNet** [comment]
Still baffling, but I think we can create a producible build pretty easily now, anyway.

**Comment by StachuDotNet** [comment]
....why do we need 16 `lets`?

**Comment by pbiggar** [comment]
Possible candidate

https://github.com/darklang/dark/blob/main/fsharp-backend/src/LibExecution/Interpreter.fs#L657

**Comment by StachuDotNet** [comment]
putting a `printfn "Hello - %A" arglist` there to see if that's the issue

**Comment by StachuDotNet** [comment]
Wait, but the problem is just in deserialization - we're not running the interpreter at all.

**Comment by pbiggar** [comment]
ah. JsonFSharpConverter?

**Comment by StachuDotNet** [comment]
Good hunch - no error if the serializer doesn't register that converter. Something to follow up on tomorrow.

**Comment by StachuDotNet** [comment]
I vaguely recall some flag existing where certain packages could be excluded from AOT compilation. Something to look into.

**Comment by pbiggar** [comment]
Maybe report the issue to FSTJ to see if they've any idea. They might be motivated to help.

**Comment by StachuDotNet** [comment]
Will do, once I have a minimal repro.

**Comment by StachuDotNet** [comment]
Maybe the easiest option: ship a Debug compilation or a non-AOT Release compilation of FSharp.SystemTextJson rather than the AOT Release one, if that truly is the problem.

**Comment by StachuDotNet** [comment]
Resolved by #4089 - seems this is specific to Ply.
I'm continuing #4071 to create a minimal repro, separately, so we can inform relevant parties.
